internal/testpeertls/revocations.go

```diff
@@ -8,50 +8,88 @@ import (
 	"crypto/x509"
 	"crypto/x509/pkix"
 
+	"storj.io/storj/pkg/identity"
 	"storj.io/storj/pkg/peertls"
 	"storj.io/storj/pkg/peertls/extensions"
-	"storj.io/storj/pkg/pkcrypto"
 )
 
 // RevokeLeaf revokes the leaf certificate in the passed chain and replaces it
 // with a "revoking" certificate, which contains a revocation extension recording
 // this action.
-func RevokeLeaf(keys []crypto.PrivateKey, chain []*x509.Certificate) ([]*x509.Certificate, pkix.Extension, error) {
-	var revocation pkix.Extension
-	revokingKey, err := pkcrypto.GeneratePrivateKey()
-	if err != nil {
-		return nil, revocation, err
+func RevokeLeaf(caKey crypto.PrivateKey, chain []*x509.Certificate) ([]*x509.Certificate, pkix.Extension, error) {
+	if len(chain) < 2 {
+		return nil, pkix.Extension{}, extensions.Error.New("revoking leaf implies a CA exists; chain too short")
+	}
+	ca := &identity.FullCertificateAuthority{
+		Key:       caKey,
+		Cert:      chain[peertls.CAIndex],
+		RestChain: chain[:peertls.CAIndex+1],
```

Comment:
Is it possible for the index to be out of bounds?

Reply:
the condition on :21 should prevent that